File: service/app/common/ApiResponse.php
<?php
/**
 * Copyright (c) 2026 erik <erik@erik.xyz> — https://erik.xyz
 */

namespace app\common;

/**
 * 统一 API 响应格式
 *
 * 成功: {"code":0, "msg":"ok", "data":{...}}
 * 失败: {"code":1, "msg":"error", "data":null}
 * 分页: {"code":0, "msg":"ok", "data":{"list":[...], "total":100, "page":1, "per_page":20}}
 */
class ApiResponse
{
    /**
     * 成功响应
     */
    public static function success(mixed $data = null, string $msg = 'ok'): \support\Response
    {
        return json([
            'code' => 0,
            'msg' => $msg,
            'data' => $data,
        ]);
    }

    /**
     * 失败响应
     */
    public static function fail(string $msg = 'error', int $code = 1, mixed $data = null): \support\Response
    {
        return json([
            'code' => $code,
            'msg' => $msg,
            'data' => $data,
        ]);
    }

    /**
     * 分页响应
     */
    public static function paginate(array $items, int $total, int $page, int $perPage): \support\Response
    {
        return json([
            'code' => 0,
            'msg' => 'ok',
            'data' => [
                'list' => $items,
                'total' => $total,
                'page' => $page,
                'per_page' => $perPage,
            ],
        ]);
    }
}

File: apps/flutter/lib/core/api/api_response.dart
class ApiResponse<T> {
  final int code;
  final String msg;
  final T? data;

  bool get isSuccess => code == 0;

  ApiResponse({required this.code, required this.msg, this.data});

  static ApiResponse<T> fromJson<T>(Map<String, dynamic> json) {
    return ApiResponse(
      code: json['code'] ?? -1,
      msg: json['msg'] ?? 'Unknown error',
      data: json['data'] as T?,
    );
  }
}

class PaginatedData<T> {
  final List<T> list;
  final int total;
  final int page;
  final int perPage;

  bool get hasMore => page * perPage < total;

  PaginatedData({
    required this.list,
    required this.total,
    required this.page,
    required this.perPage,
  });

  factory PaginatedData.fromJson(Map<String, dynamic> json, T Function(Map<String, dynamic>) fromJsonT) {
    return PaginatedData(
      list: (json['list'] as List).map((e) => fromJsonT(e as Map<String, dynamic>)).toList(),
      total: json['total'] ?? 0,
      page: json['page'] ?? 1,
      perPage: json['per_page'] ?? 20,
    );
  }
}
